tests/lib/fortigatelocalin_test.py

```diff
@@ -0,0 +1,496 @@
+import string
+import unittest
+import mock
+
+from capirca.lib import fortigate
+from capirca.lib import nacaddr
+from capirca.lib import naming
+from capirca.lib import policy
+
+
+GOOD_HEADER = """
+header {
+  comment:: "this is a test acl"
+  target:: fortigate from-id 2
+}
+"""
+
+GOOD_HEADER_1 = """
+header {
+  comment:: "this is a test acl"
+  target:: fortigate ngfw-mode policy-based
+}
+"""
+
+BAD_HEADER = """
+header {
+  comment:: "this is a test acl"
+  target:: fortigate edge-filter
+}
+"""
+
+TERM_TEMPLATE = """
+term good-term-2 {{
+  source-interface:: {src_interface}
+  destination-interface:: {dest_interface}
+  protocol:: {protocol}
+  destination-address:: {dest_addr}
+  destination-port:: {dest_port}
+  source-address:: {src_addr}
+  source-port:: {src_port}
+  action:: {action}
+  logging:: {logging}
+}}
+"""
+
+_SP = '    '
+
+EXP_INFO = 2
+
+
+class CustomFormatter(string.Formatter):
+  """
+  Checks the custom formatter for fortigate output.
+
+  """
+  DEFAULT_VALUES = {
+      'src_interface': 'wan1',
+      'dest_interface': 'wan2',
+      'protocol': 'tcp',
+      'src_addr': 'SOME_HOST',
+      'dest_addr': 'SOME_HOST',
+      'src_port': 'HTTP',
+      'dest_port': 'HTTP',
+      'action': 'accept',
+      'logging': 'true'
+  }
+
+  def format(*args, **kwargs):
+    if 'remove_fields' in kwargs or 'add_fields' in kwargs:
+      args = list(args)
+
+    if 'remove_fields' in kwargs:
+      for field in kwargs['remove_fields']:
+        remove_regex = '.*' + field + '.*'
+        args[1] = re.sub(remove_regex, '', args[1])
+
+    if 'add_fields' in kwargs:
+      add_fields_string = ""
+      for field, value in kwargs['add_fields'].items():
+        add_fields_string += "  " + field + ":: " + value + "\n"
+      args[1] = args[1][:-3] + add_fields_string + args[1][-3:]
+
+    return string.Formatter.format(*args, **kwargs)
+
+  def get_value(self, key, args, kwds):
+    try:
+      return kwds[key]
+    except KeyError:
+      return self.DEFAULT_VALUES[key]
+
+
+class FortigateTest(unittest.TestCase):
+  """
+  Fortigate test class.
+
+  """
+  def setUp(self):
+    self.naming = mock.create_autospec(naming.Naming)
+
+    def get_addr_side_eff(host):
+      hosts = {
+          'SOME_HOST': [nacaddr.IP('10.0.0.0/8')],
+          'SOME_HOST2': [nacaddr.IP('20.0.0.0/8')],
+          'SOME_HOST6': [nacaddr.IP('fec0::/10')]
+      }
+      return hosts[host]
+
+    def get_port_side_eff(*args):
+      hosts = {
+          'HTTP': ['80'],
+          'HTTPS': ['443'],
+          'SSH': ['22'],
+          'WHOIS': ['43']
+      }
+      return hosts[args[0]]
+
+    self.naming.GetNetAddr.side_effect = get_addr_side_eff
+    self.naming.GetServiceByProto.side_effect = get_port_side_eff
+    self.fmt = CustomFormatter()
+
+  def testGoodHeader(self):
+    """
+    Tests a good header value.
+
+    """
+    term = self.fmt.format(TERM_TEMPLATE)
+    acl = fortigate.Fortigate(policy.ParsePolicy(GOOD_HEADER + term,
```

Comment:
Should we be calling the ForitgateLocalIn classes here in this entire file - so if other changes get introduced over time, we continue to test the right platform.